getting_started/workflow/best_practices/autoloads_versus_internal_nodes.rst

```diff
@@ -1,103 +1,95 @@
 .. _doc_autoloads_versus_internal_nodes:
 
-Autoloads versus internal nodes
+Autoloads versus regular nodes
+==============================
+
+Godot offers a feature to automatically load nodes at the root of your project,
+making allowing you to access them globally, that can fulfill the role of a
+Singleton: :ref:`doc_singletons_autoload`. These auto-loaded nodes are not freed
+when you change the scene from code with ``SceneTree.change_scene``.
+
+In this guide, you will learn when to use the Autoload feature, and techniques
+you can use to avoid it.
+
+The cutting audio issue
+=======================
+
+Other engines can encourage the use of creating manager classes, singletons that
+organize a lot of functionality into a globally accessible object. Godot offers
+many ways to avoid global state thanks to the node tree and signals.
+
+For example, let's say we are building a platformer and want to collect coins
+that play a sound effect. There's a node for that: the :ref:`AudioStreamPlayer
+<class_AudioStreamPlayer>`. But if we if call the ``AudioStreamPlayer`` while it
+is already playing a sound, the new sound interrupts the first.
+
+A solution is to code a global, auto-loaded sound manager class. It generates a
+pool of ``AudioStreamPlayer`` that cycle through as each new request for sound
+effects comes in. Say we call that class ``Sound``, you can use it from anywhere
+in your project by calling ``Sound.play("coin_pickup.ogg")``. This solves the
+problem in the short term but causes more problems:
+
+1. **Global state**: one object is now responsible for all objects' data. If the
+   ``Sound`` class has errors or doesn't have an AudioStreamPlayer available,
+   all the nodes calling it can break.
+
+2. **Global access**: now that any object can call ``Sound.play(sound_path)``
+   from anywhere, there's no longer an easy way to find the source of a bug.
```

Comment:
> there's no longer an easy way to find the source of a bug.

I think that we should elaborate here on what we mean. *How* does this make it harder to find bugs? What could potentially happen and what could we have trouble doing in our attempt to solve it?

Reply:
I took this from the original article. I wouldn't use a singleton for things like these so I don't have examples from experience but I'd guess that:

1. You have to navigate to the Sound class to change the code
2. If you run out of streams and the Sound class cycles and e.g. replaces a long audio sample with a new one, it's not that easy to figure out why that's happening in the middle of a project, especially as you won't get a backtrace and when you try the scene in isolation with F6, everything may work fine.

Does that seem good enough of an example? In my exp. the issues of singletons are not obvious at first, they really start to hurt when your project grows and you have a hard dependency in many places at once. It's difficult to give an example of that.

Reply:
Well, those examples work well too. But my primary concern with things like this is that the location of a problem, namely the context in which the method is called, is no longer clear when the functionality can be accessed from anywhere. Rather than dealing with a subset of nodes internal to a scene, your potential domain of trouble-making code spans the entire project. Any code anywhere could have passed erroneous data to the `Sound` singleton. When code is kept inside a scene, there is no realy concern for this problem as only a handful of scripts are usually involved total, let alone dealing with the topic that the singleton handles.